src/internal/observable/generate.ts

```diff
@@ -360,24 +360,23 @@ export function generate<T, S>(
     }
   }
 
-  return new Observable<T>((subscriber) => {
-    let state = initialState;
-    if (scheduler) {
-      let needIterate = false;
-      caughtSchedule(subscriber, scheduler, function () {
-        needIterate ? (state = iterate!(state)) : (needIterate = true);
-        condition && !condition(state) ? subscriber.complete() : subscriber.next(resultSelector(state));
-        this.schedule();
-      });
-    } else {
-      do {
-        if (condition && !condition(state)) {
-          subscriber.complete();
-        } else {
-          subscriber.next(resultSelector(state));
-          !subscriber.closed && (state = iterate!(state));
-        }
-      } while (!subscriber.closed);
+  // The actual generator used to "generate" values.
+  function* gen() {
+    for (let state = initialState; !condition || condition(state); state = iterate!(state)) {
+      yield resultSelector(state);
     }
-  });
+  }
+
+  // TODO: Figure out what is going on with the types here!
+
+  // We use `defer` because we want to defer the creation of the iterator from the iterable.
+  return defer(
+    scheduler
+      ? // If a scheduler was provided, use `scheduleIterable` to ensure that iteration/generation
+        // happens on the scheduler.
+        () => scheduleIterable(gen(), scheduler!)
+      : // Otherwise, if there's no scheduler, we can just use the generator function directly in
+        // `defer` and executing it will return the generator (which is iterable).
+        (gen as any)
+  ) as any;
```

Comment:
issue: The problem with the types is that the ternary introduces a union type for the callback received by `defer`. more simply:

```ts
let b = false;
defer(b ? () => Promise.resolve(1) : [2]);
```

Here, the type of the callback is `(() => Promise<number) | (() => [number])` and that union defeats `defer`'s type signature's inference.

You could solve the problem with a narrower assertion (than `any`), like this:

```ts
  return defer(
    (scheduler
      ? // If a scheduler was provided, use `scheduleIterable` to ensure that iteration/generation
        // happens on the scheduler.
        () => scheduleIterable(gen(), scheduler!)
      : // Otherwise, if there's no scheduler, we can just use the generator function directly in
        // `defer` and executing it will return the generator (which is iterable).
        gen) as () => ObservableInput<T>
  );
```

Or could move `defer` into the ternary, like this:

```ts
  return scheduler
    ? // If a scheduler was provided, use `scheduleIterable` to ensure that iteration/generation
      // happens on the scheduler.
      defer(() => scheduleIterable(gen(), scheduler!))
    : // Otherwise, if there's no scheduler, we can just use the generator function directly in
      // `defer` and executing it will return the generator (which is iterable).
      defer(gen)
  );
```

I would opt for the former - the narrower assertion.